### PR TITLE
feat: EDV document processor implementation

### LINKED ISSUE #2249: Add concrete EDV document processor (encrypt/decrypt) implementation

### PR DESCRIPTION
**Pull request by DRK3**
Added a concrete document processor implementation. It can encrypt and decrypt between structured documents and encrypted documents.

Signed-off-by: Derek Trider <Derek.Trider@securekey.com>

closes #2249

### CODE REVIEW

File: pkg/storage/edv/documentprocessor/documentprocessor.go [hunk 1]
```diff
@@ -7,12 +7,91 @@ SPDX-License-Identifier: Apache-2.0
...
+	failDeserializeJWE              = "failed to deserialize JWE: %w"
+	failDecryptJWE                  = "failed to decrypt JWE: %w"
+	failUnmarshalStructuredDocument = "failed to unmarshal structured document: %w"
+)
+
+type marshalFunc func(interface{}) ([]byte, error)
+
 // DocumentProcessor represents a type that can encrypt and decrypt between
 // Structured Documents and Encrypted Documents.
 type DocumentProcessor interface {
```
Comment (troyronda) [inline pkg/storage/edv/documentprocessor/documentprocessor.go:30]: Where is this interface consumed?

The DocumentProcessor shouldn't be coupled to EDV so you should consider another package for it.

File: pkg/storage/edv/documentprocessor/documentprocessor.go [hunk 2]
```diff
@@ -7,12 +7,91 @@ SPDX-License-Identifier: Apache-2.0
...
 	Encrypt(*edv.StructuredDocument) (*edv.EncryptedDocument, error)
 	Decrypt(*edv.EncryptedDocument) (*edv.StructuredDocument, error)
 }
+
+// AriesDocumentProcessor uses Aries crypto to encrypt and decrypt between
+// Structured Documents and Encrypted Documents.
+type AriesDocumentProcessor struct {
```
Comment (troyronda) [inline pkg/storage/edv/documentprocessor/documentprocessor.go:37]: Shouldn't have the Aries prefix. Fixable if you don't put the interface and concrete implementation together.

**Review by sudeshrshetty — APPROVED** [review]

**Review by aholovko — APPROVED** [review]

File: pkg/storage/edv/documentprocessor/documentprocessor.go (hunk 1 above)
Comment (DRK3) [reply pkg/storage/edv/documentprocessor/documentprocessor.go:30]: This is used by the (EDV) `FormatProvider` for doing encryption/decryption before calling the underlying store. This `DocumentProcessor` and the `FormatProvider` together are what allows me to decouple the EDV REST API from the underlying store.

It's specifically for the EDV Structured Documents and Encrypted Document models as defined in the spec. It's not for general purpose encryption/decryption. Can you clarify what you mean by decoupling it from EDV?

Comment (troyronda) [reply pkg/storage/edv/documentprocessor/documentprocessor.go:30]: You shouldn’t normally be exporting both an interface and the concrete implementation from the same package.

File: pkg/storage/edv/documentprocessor/documentprocessor.go (hunk 2 above)
Comment (DRK3) [reply pkg/storage/edv/documentprocessor/documentprocessor.go]: Fixed.

File: pkg/storage/edv/documentprocessor/documentprocessor.go (hunk 1 above)
Comment (DRK3) [reply pkg/storage/edv/documentprocessor/documentprocessor.go]: Fixed.

**Review by troyronda — APPROVED** [review]

**Review by fqutishat — APPROVED** [review]

**Pull request merged**